cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -285,15 +287,56 @@ multiReplDecision ctx compiler flags =
 -- For more details on how this works, see the module
 -- "Distribution.Client.ProjectOrchestration"
 replAction :: NixStyleFlags ReplFlags -> [String] -> GlobalFlags -> IO ()
-replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings globalFlags =
-  withContextAndSelectors verbosity AcceptNoTargets (Just LibKind) flags targetStrings globalFlags ReplCommand $ \targetCtx ctx targetSelectors -> do
+replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings' globalFlags = do
+  -- NOTE: The REPL will work with no targets in the context of a project if a
+  -- single package is in the same directory as the project file. To have the
+  -- same behaviour when the package is somewhere else we adjust the targets.
+  targetStrings <-
+    if null targetStrings'
+      then withCtx silent targetStrings' $ \targetCtx ctx _ ->
+        return . fromMaybe [] $ case targetCtx of
+          ProjectContext ->
+            let pkgs = projectPackages $ projectConfig ctx
+             in if length pkgs == 1
+                  then pure <$> listToMaybe pkgs
+                  else Nothing
+          _ -> Nothing
+      else return targetStrings'
```

Comment:
This looks ad hoc. I'd rather we just fail and make people specify target rather than feedling with context like that. No one really requested that the default behaviour with a single package in CWD was extended to a single package anywhere, right?

Reply:
I've removed this target string manipulation.